include/picongpu/plugins/xrayScattering/XrayScattering.hpp

```diff
@@ -0,0 +1,846 @@
+                    amplitudeReceive.assign(
+                        bufferSize,
+                        complex_X( 0.0 )
+                    );
+                }
+                // Allocate amplitude buffer.
+                amplitude = pmacc::memory::makeUnique< ComplexBuffer >(
+                    DataSpace< DIM1 >( bufferSize ) );
+                // Initialize, on device, its fields with zero.
+                amplitude->getDeviceBuffer( ).setValue( 0.0 );
+
+                // Go to PIC unit system.
+                q_min = q_min * 1.0e10 * UNIT_LENGTH;
+                q_max = q_max * 1.0e10 * UNIT_LENGTH;
+                // Set the q-space grid spacing.
+                q_step = ( q_max - q_min ) /
+                         precisionCast< float_X >( numVectors );
+
+                // Rank 0 creates the output directory.
+                pmacc::Filesystem< simDim > & fs =
+                    Environment< simDim >::get( ).Filesystem( );
+                if ( isMaster )
+                {
+                    fs.createDirectory( "xrayScatteringOutput" );
+                    fs.setDirectoryPermissions( "xrayScatteringOutput" );
+                }
+
+                // Chose the solver for populating a TmpField with the electron
+                // density (either the species density or the bound electron
+                // density).
+                using ElectronDensitySolver =
+                    typename DetermineElectronDensitySolver<
+                    T_ParticlesType >::type;
+                // Output unit:
+                const float_64 amplitudeUnit = static_cast< float_64 >(
+                    FieldTmp::getUnit< ElectronDensitySolver >( )[ 0 ] ) *
+                    CELL_WIDTH_SI * CELL_HEIGHT_SI * CELL_DEPTH_SI
+                    * ELECTRON_RADIUS_SI;
+
+                // Set the total number of cells in the simulation.
+                totalSimulationCells =  Environment< simDim >::get( ).SubGrid(
+                    ).getGlobalDomain( ).size.productOfComponents( );
+
+                // Initialize an object responsible for output writing.
+                dataWriter = pmacc::memory::makeUnique< XrayScatteringWriter<
+                    float_X > >(
+                        pluginPrefix + "Output",
+                        fileExtension,
+                        "xrayScatteringOutput",
+                        outputLayout,
+                        compressionMethod,
+                        precisionCast< uint64_t >( numVectors ),
+                        q_step,
+                        amplitudeUnit,
+                        totalSimulationCells
+                    );
+                // Set the output period.
+                outputPeriod = pluginSystem::toTimeSlice( outputPeriod_s );
+            }
+        }
+
+
+        void pluginUnload( ) override
+        {
+            // if ( !notifyPeriod.empty( ) )
+            // {
+            //     CUDA_CHECK( cudaGetLastError( ) );
+            // }
```

Comment:
Huh, good question. It obviously works without this code. This is something I took over from the previous PR. @sbastrakov I forgot if and why this would be necessary, can I remove it?

Reply:
I think since you don't do any operations here, having this check does not add anything, could just keep the body empty. (Also `cudaGetLastError` is now `cuplaGetLastError` in `dev`)